### PR TITLE
driver: add Next method for Recordset and use it in driver

### PR DESCRIPTION
**Pull request by coocood**
So we can implement `driver.Rows.Next` without a goroutine.

### CODE REVIEW

File: rset/rset.go [hunk 1]
```diff
@@ -32,4 +35,11 @@ type Recordset interface {
 
 	// Get rows data by using limit/offset.
 	Rows(limit, offset int) (rows [][]interface{}, err error)
```
Comment (siddontang) [inline rset/rset.go]: Row gets rows data

File: rset/rsets/rsets.go [hunk 2]
```diff
@@ -77,27 +77,35 @@ func (r Recordset) FirstRow() (row []interface{}, err error) {
...
-		case limit < 0:
-			rows = append(rows, row)
-			return true, nil
-		case limit == 0:
-			return false, nil
-		default: // limit > 0
-			rows = append(rows, row)
+		offset--
+	}
+	for limit != 0 {
```
Comment (ngaut) [inline rset/rsets/rsets.go]: Need comments when limit is -1 and 0

File: rset/rsets/rsets.go [hunk 3]
```diff
@@ -77,27 +77,35 @@ func (r Recordset) FirstRow() (row []interface{}, err error) {
 // Rows implements rset.Recordset.
 func (r Recordset) Rows(limit, offset int) ([][]interface{}, error) {
 	var rows [][]interface{}
-	err := r.Do(func(row []interface{}) (bool, error) {
-		if offset > 0 {
-			offset--
-			return true, nil
+	defer r.Plan.Close()
+	for offset > 0 {
```
Comment (ngaut) [inline rset/rsets/rsets.go]: Comment: // Move to offset

File: driver.go [hunk 4]
```diff
@@ -352,54 +351,17 @@ func (r *driverResult) RowsAffected() (int64, error) {
...
-		}); e != nil {
-			err = e
-		}
-
-		select {
-		case r.rows <- err:
-		case <-r.done:
-		}
-	}()
+		rs: rs,
```
Comment (qiuyesuifeng) [inline driver.go]: How about directly `return r := &{xxx}`

File: driver.go [hunk 5]
```diff
@@ -352,54 +351,17 @@ func (r *driverResult) RowsAffected() (int64, error) {
...
+	rs rset.Recordset
 }
 
 func newEmptyDriverRows() *driverRows {
-	r := &driverRows{
-		done: make(chan int, 1),
-	}
-	r.done <- 1
-	return r
+	return &driverRows{}
```
Comment (qiuyesuifeng) [inline driver.go]: Seems no use to extract this function?

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

File: expression/expressions/subquery.go [hunk 6]
```diff
@@ -62,27 +62,27 @@ func (sq *SubQuery) Eval(ctx context.Context, args map[interface{}]interface{})
...
-		return true, nil
-	})
-
+	row, err := p.Next(ctx)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-
+	if row == nil {
+		return sq.Value, nil
```
Comment (qiuyesuifeng) [inline expression/expressions/subquery.go]: Can `return nil, nil` here?
That seems clearer.

**Pull request merged**